# Stochastic pulsing of gene expression enables the generation of spatial patterns in Bacillus subtilis biofilms

Stochastic pulsing of gene expression can generate phenotypic diversity in a genetically identical population of cells, but it is unclear whether it has a role in the development of multicellular systems. Here, we show how stochastic pulsing of gene expression enables spatial patterns to form in a model multicellular system, Bacillus subtilis bacterial biofilms. We use quantitative microscopy and time-lapse imaging to observe pulses in the activity of the general stress response sigma factor σB in individual cells during biofilm development. Both σB and sporulation activity increase in a gradient, peaking at the top of the biofilm, even though σB represses sporulation. As predicted by a simple mathematical model, increasing σB expression shifts the peak of sporulation to the middle of the biofilm. Our results demonstrate how stochastic pulsing of gene expression can play a key role in pattern formation during biofilm development.

## Repository explanation
The code in this repository assumes that you have 
a directory called `datasets` in your local repository.
The contents of this directory can be found at https://dx.doi.org/10.5281/zenodo.3544513/. 

There is also an assumed directory called `proc_data` that holds the images to be analysed. Please contact the authors for this data. 

### Segmentation Code
The main segmentation functions used are found in the functions
* `basic_segment` in `lib/processing/slice10x/segment.py`.
* `laphat_segment_v1` in file `lib/cell_segmenter.py`.


### Model 
Information about the computational model are found in [model/README.md](model/README.md).

### Workflow files
In `workflows/` are bash shell scripts that show how raw images were pushed through the pipelines. 
They were not meant to be run as a standard script but rather interactively like a notebook running different sections. They should be used more as a record of parameters and what flags to pass the different tools, that a full stand alone pipeline. 

## Figures

### Figure 1
Single cells from "pad movies" were segmented and tracked using the [Schnitzcells Matlab package](http://easerver.caltech.edu/wordpress/schnitzcells/). 
The main files to generate the data for Figure 1 are `matlab_padmovies/fig1_generate_hist_data.m` and `matlab_padmovies/fig1_generate_track_data.m`. 
The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `padmovies_brightfield.zip`.

From here you can generate Figure 1 by running 
```
python figures/figure_padmovies/padmovie_trace_main.py
```


### Figure 2
Biofilm gradients. 

The file `workflows/lsm780_sigbyfp.sh` shows how various tools were used to segment and extract gradient data. 

* The biofilm segmentation tool is `bin/segment_10x.py`. 
* The biofilm segmentation algorithm is implemented in function `basic_segment` in `lib/processing/slice10x/segment.py`.
* Gradients were computed using `bin/gradient_10x_maker.py` 

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM780_10x_sigb.zip`.

* The mean gradient for each strain was computed using `analysis/summarise_10x_gradients.py`
* Figure generated with `figures/figure_sigb_10x_grad/fig_10x_grad_rot.py` 


### Figure 3
Single cell histograms. 

The file `workflows/lsm700_sigbyfp.sh` shows how the various parts of the analysis pipeline were strung together. 

* The cell segmentation and quantification tool is `bin/giant63_split_cellquant.py`
* The cell segmentation algorithm is implemented in function `laphat_segment_v1` in file `lib/cell_segmenter.py`.

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM700_63x_sigb.zip`.

The figure was generated by running
```
python figures/figure_63x_sigb_histo/make_histograms_quad.py
```

### Figure 4
This figure shows the results of our live biofilm movies. 

Gradient data (B) was processed using the workflow `workflows/iphox_live_tiles.sh`. 
The images were segmented using the same code as used for Figure 2. 

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `iphox_gradient_snaps.zip`.

The single cell data was processed using the workflow `workflows/iphox_movies.sh`. 
The cells were manually tracked using a custom interface `bin/manual_tracker_sql.py`. 

The figure was generated by running
```
python figures/figure_bf_movie/figure_bf_movie.py
```

### Figure 5
This figure compares σB gradients and spore gradients.
Also plots σB activity vs sporulation activity. 

Single cell data (see Figure 2), Spore gradient data (see Figure 7). 
These data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the files `LSM700_63x_sigb.zip` and `LSM700_63x_sspb_giant.zip`.

Single cell three reporter data were processed as in the workflow `workflows/workflow_3fp.sh`.
These large tilescan images were shrunk down to 1/20 their original size and segmented using the method from Figure 2 and manually corrected. 
The small mask was increased back up to full size and the images were processed using the single cell segmentation algorithm from Figure 3. 

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM780_63x_spoiid_v_sigb.zip`.

The figure was generated with
```
python figures/figure_sigb_repress_spores/sigb_vs_spore.py
```

### Figure 6
Figure shows results of our computational model. 
For more details see [model/README.md](model/README.md). 

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `model_results.zip`.

The figure was generated with 
```
python figures/figure_model_summary.py
```

### Figure 7
This figure shows σB gradients and spore gradients for WT and the 2xQP mutant. 

The σB gradient data is from the Figure 2 dataset and is found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM780_10x_sigb.zip`.

The spore gradients were extracted from large tilescan images using the workflow `workflows/workflow_giantsspb.sh`.
The biofilm was segmented using the same method as Figure 2 and manually adjusted using the mask editor `bin/mask_maker.py --minidraw`. 

The data was summarised for quick plotting using `analysis/summarize_cell_spore_counts.py`.

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM700_63x_sspb_giant.zip`.

The figure was generated using
```
 python figures/figure_allspore_2xqp_combo/figure_all_spore_sigb_combo.py
```

## Supplemental Figures

### Supplemental Figure 1 
Figure showing that biofilms from our live movie assay look like normal biofilms. 

`pdflatex figures/sup_moviebf/sup_moviebf.tex`

### Supplemental Figure 2
Figure showing σB activity gradients over time for WT and two knockout strains. 
The data is the same as used for Figure 2 and can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM700_63x_sigb.zip`.

Gradients are computed from single cell data using `analysis/summarise_63x_gradients.py`.

Figure created with `python figures/sup_63x_gradients/sup_63x_grad_error.py`

### Supplemental Figure 3
Figure shows other sigma factors do not have gradients and its not a side effect of YFP. 

Figure created with `python figures/sup_notjustYFP/sup_notjustYFP.py`

### Supplemental Figure 4
Figure showing σA activity does not show gradients or hetrogenious expression if we use a YFP reporter.
The data is the same as used for Sup Figure 8 and can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `new_strain_snaps1.zip`.

Figure created with `fig_sigayfp_3panels_histogram.py`

### Supplemental Figure 5
Figure showing σB activity gradient in biofilms growing on roots. 

Figure created with `python figures/sup_roots/sup_roots.py`

### Supplemental Figure 6
Figure shows σB activity histograms at different depths in biofilms from WT and mutant strains. 
The dataset is the same as Figure 3 and can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM700_63x_sigb.zip`.

The figure was generated using `python figures/figure_63x_sigb_histo/joy_plots_of_gradients.py`.

### Supplemental Figure 7
Single cell trace data from biofilm movies. 
The data is the same as from Figure 4 and can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `iphox_gradient_snaps.zip`.

The figure was generated using `python figures/supfig_bfmovie_pulses/supfig_bfmovie_pulse.py`.


### Supplemental Figure 8
The figure shows biofilm cross sections, single cell snapshots, and expression histograms for a number of mutants.

The workflow file `workflows/lsm700_newstrains.sh` shows the tools used to process the raw images.
The process is identical to that for Figure 3. 
The same segmentation code is used. 

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `new_strain_snaps1.zip`.

The figure was created using `python figures/sup_transformable_live/fig_transformable.py`.

### Supplemental Figure 9
The figure shows two 2d-histograms showing the correlation between σB and an early sporulation gene. 

The data is the same as used in Figure 5.C and can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM780_63x_spoiid_v_sigb.zip`.

The figure was generated using `python figures/supfig_spoiid_anticorr/supfig_spoiid_anticorr.py`

### Supplemental Figure 10
The figure shows spore gradients and the distribution of the peaks of the sporulation gradient. 

The data is the same as used in Figure 7 and can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM700_63x_sspb_giant.zip`.

The figure was generated with `python figures/figure_allspore_2xqp_combo/figure_spore_counts.py`


### Supplemental Figure 11
2xQP gradients and σB activity at different depths histograms. 
This data is processed as in Figure 3, Sup Fig 2, Sup Fig 6.

The resulting data can be found at [https://dx.doi.org/10.5281/zenodo.3544513/](https://dx.doi.org/10.5281/zenodo.3544513/) in the file `LSM700_63x_sigb.zip`.

`python figures/sup_2xqp/sup_2xqp.py` 

### Supplemental Figure 12
Figure showing a long section of spore images.

Figure created with `python figures/sup_longspore_im/sup_longspore.py`

